controllers/mattermost/clusterinstallation/migration.go

```diff
@@ -0,0 +1,264 @@
+package clusterinstallation
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	mattermostv1alpha1 "github.com/mattermost/mattermost-operator/apis/mattermost/v1alpha1"
+	mmv1beta "github.com/mattermost/mattermost-operator/apis/mattermost/v1beta1"
+	"github.com/mattermost/mattermost-operator/pkg/mattermost/healthcheck"
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type MigrationResult struct {
+	Status    string
+	RequeueIn time.Duration
+	Finished  bool
+}
+
+// HandleMigration performs necessary steps to migrate ClusterInstallation to Mattermost.
+func (r *ClusterInstallationReconciler) HandleMigration(ci *mattermostv1alpha1.ClusterInstallation, logger logr.Logger) (MigrationResult, error) {
+	logger = logger.WithValues("migration", "v1beta")
+
+	err := r.IsConvertible(ci)
+	if err != nil {
+		logger.Error(err, "ClusterInstallation cannot be converted to Mattermost CR")
+		return MigrationResult{
+			Status: fmt.Sprintf("Migration to Mattermost cannot be performed safely: %s", err.Error()),
+		}, nil
+	}
+
+	name := types.NamespacedName{Name: ci.Name, Namespace: ci.Namespace}
+
+	var existingMM mmv1beta.Mattermost
+	err = r.Get(context.Background(), name, &existingMM)
+	if err != nil && !k8sErrors.IsNotFound(err) {
+		return MigrationResult{}, errors.Wrap(err, "failed to check if Mattermost CR exists")
+	}
+	if k8sErrors.IsNotFound(err) {
+		finished := false
+		finished, err = r.initializeMigration(ci, logger)
+		if err != nil {
+			return MigrationResult{}, errors.Wrap(err, "failed to initialize migration")
+		}
+
+		if !finished {
+			return MigrationResult{
+				RequeueIn: 10 * time.Second,
+				Status:    "Migration to Mattermost is in progress - recreating deployment",
+			}, nil
+		}
+	}
+
+	if existingMM.Status.State != mmv1beta.Stable {
+		logger.Info("Migration not finished. Waiting for Mattermost to be in 'stable' state")
+		return MigrationResult{
+			RequeueIn: 10 * time.Second,
+			Status:    "Migration to Mattermost is in progress - waiting for Mattermost to be ready",
+		}, nil
+	}
+
+	logger.Info("Migration finished. Removing old Replica Sets and ClusterInstallation")
+	err = r.cleanupReplicaSets(ci)
+	if err != nil {
+		return MigrationResult{}, errors.Wrap(err, "failed to cleanup old Replica Sets")
+	}
```

Comment:
I am curious if this is needed due to this being a migration and there are differences from the normal ClusterInstallation resource deletion which happens immediately after this? This is normally handled for us when that deletion occurs, but I am assuming there is a difference here with the migration logic underway.

Reply:
Yes, that is due to difference in how we are handling the migration. 
To avoid downtime we need to delete old Deployment orphaning its Replica Sets so that Pods keep running until we create new Deployment (this is needed as we change the immutable `selector` field). 
In usual scenario the Replica Sets have Owner References to the Deployment and Deployment to the ClusterInstallation therefor removing ClusterInstallation cleans out the whole thing. However if we remove deployment earlier and orphan Replica Sets, there is no longer Owner Reference that triggers their removal so we need to clean the up manually.